src/preferences/upgrade.cpp

```diff
@@ -555,34 +590,24 @@ UserSettingsPointer Upgrade::versionUpgrade(const QString& settingsPath) {
                     configVersion.startsWith("2.5.0-"))) {
         // Proactively move users to an all-shader waveform widget type and set the
         // framerate to 60 fps
-        bool okType = false, okBackend = false, okOptions = false;
-        auto waveformType =
-                config->getValueString(ConfigKey("[Waveform]", "WaveformType"))
-                        .toInt(&okType);
-        auto waveformBackend =
-                config->getValueString(ConfigKey("[Waveform]", "use_hardware_acceleration"))
-                        .toInt(&okBackend);
-        auto waveformOption =
-                config->getValueString(ConfigKey("[Waveform]", "waveform_options"))
-                        .toInt(&okOptions);
-        if (okType && (!okBackend || !okOptions)) {
-            upgradeToAllShaders(&waveformType,
-                    reinterpret_cast<WaveformWidgetBackend*>(&waveformBackend),
-                    reinterpret_cast<
-                            allshader::WaveformRendererSignalBase::Options*>(
-                            &waveformBackend));
-            config->setValue(ConfigKey("[Waveform]", "WaveformType"),
-                    waveformType);
-            config->setValue(ConfigKey("[Waveform]", "use_hardware_acceleration"),
-                    waveformBackend);
-            config->setValue(ConfigKey("[Waveform]", "waveform_options"),
-                    waveformOption);
-        }
-        // Only update the framerate if the version we are migrating from is less than 2.4.0.
-        if (QVersionNumber::fromString(configVersion) < QVersionNumber(2, 4, 0)) {
-            config->set(ConfigKey("[Waveform]", "FrameRate"), ConfigValue(60));
-        }
-
+        int waveformType =
+                config->getValue<int>(ConfigKey("[Waveform]", "WaveformType"));
+        // values might be out of range for the enum, avoid undefined
+        // behavior by not casting to the enum type just yet.
+        int waveformBackend = config->getValue<int>(
+                ConfigKey("[Waveform]", "use_hardware_acceleration"));
+        int waveformOption = config->getValue<int>(ConfigKey("[Waveform]", "waveform_options"));
+        auto [correctedWaveformType,
+                correctedWaveformBacked,
+                correctedWaveformOption] =
+                upgradeToAllShaders(
+                        waveformType, waveformBackend, waveformOption);
+        config->setValue(ConfigKey("[Waveform]", "WaveformType"),
+                correctedWaveformType);
+        config->setValue(ConfigKey("[Waveform]", "use_hardware_acceleration"),
+                correctedWaveformBacked);
+        config->setValue<int>(ConfigKey("[Waveform]", "waveform_options"),
+                correctedWaveformOption);
```

Comment:
very clean IMO :heart_eyes: 
A little redundant, but it doesn't get much more elegant in C++ I think.